src/osd/pg_scrubber.cc

```diff
@@ -0,0 +1,2147 @@
+  omap_stats = (const struct omap_stat_t){0};
+
+  run_callbacks();
+  inconsistent_.clear();
+  missing_.clear();
+  authoritative_.clear();
+  num_digest_updates_pending = 0;
+  replica_scrubmap = ScrubMap{};
+  replica_scrubmap_pos.reset();
+
+  cleaned_meta_map = ScrubMap{};
+  needs_sleep_ = true;
+  sleep_started_at_ = utime_t{};
+
+  active_ = false;
+  pg_->publish_stats_to_osd();
+}
+
+
+/*
+ * note: performs run_callbacks()
+ * note: reservations-related variables are not reset here
+ */
+void PgScrubber::reset_internal_state()
+{
+  dout(7) << __func__ << dendl;
+
+  preemption_data.reset();
+  waiting_on_whom.clear();
+  received_maps.clear();
+
+  start_ = hobject_t{};
+  end_ = hobject_t{};
+  max_end = hobject_t{};
+  subset_last_update_ = eversion_t{};
+  shallow_errors_ = 0;
+  deep_errors = 0;
+  fixed_count_ = 0;
+  omap_stats = (const struct omap_stat_t){0};
+
+  run_callbacks();
+
+  inconsistent_.clear();
+  missing_.clear();
+  authoritative_.clear();
+  num_digest_updates_pending = 0;
+  primary_scrubmap_ = ScrubMap{};
+  primary_scrubmap_pos_.reset();
+  replica_scrubmap = ScrubMap{};
+  replica_scrubmap_pos.reset();
+  cleaned_meta_map = ScrubMap{};
+  needs_sleep_ = true;
+  sleep_started_at_ = utime_t{};
+
+  flags_ = scrub_flags_t{};
+
+  active_ = false;
+}
+
+const OSDMapRef& PgScrubber::get_osdmap() const
+{
+  return pg_->get_osdmap();  // RRR understand why we cannot use curmap_
+}
+
+ostream& operator<<(ostream& out, const PgScrubber& scrubber)
+{
+  return out << scrubber.flags_;
+}
+
+ostream& PgScrubber::show(ostream& out) const
+{
+  return out << " [ " << pg_id_ << ": " << /*for now*/ flags_ << " ] ";
+}
+
+// ///////////////////// preemption_data_t //////////////////////////////////
+
+PgScrubber::preemption_data_t::preemption_data_t(PG* pg) : pg_{pg}
+{
+  left_ = static_cast<int>(
+    pg_->get_cct()->_conf.get_val<uint64_t>("osd_scrub_max_preemptions"));
+}
+
+void PgScrubber::preemption_data_t::reset()
+{
+  std::lock_guard<std::mutex> lk{preemption_lock_};
+
+  preemptable_ = false;
+  preempted_ = false;
+  left_ =
+    static_cast<int>(pg_->cct->_conf.get_val<uint64_t>("osd_scrub_max_preemptions"));
+  size_divisor_ = 1;
+}
+
+
+// ///////////////////// ReplicaReservations //////////////////////////////////
+namespace Scrub {
+
+void ReplicaReservations::release_replica(pg_shard_t peer, epoch_t epoch)
+{
+  dout(7) << __func__ << " <ReplicaReservations> release-> " << peer << dendl;
+
+  Message* m = new MOSDScrubReserve(spg_t(pg_->info.pgid.pgid, peer.shard), epoch,
+				    MOSDScrubReserve::RELEASE, pg_->pg_whoami);
+  osds_->send_message_osd_cluster(peer.osd, m, epoch);
+}
+
+ReplicaReservations::ReplicaReservations(PG* pg, PgScrubber* scrubber, pg_shard_t whoami)
+    : pg_{pg}
+    , scrubber_{scrubber}
+    , acting_set_{pg->get_actingset()}
+    , osds_{static_cast<OSDService*>(pg_->osd)}
+{
+  epoch_t epoch = pg_->get_osdmap_epoch();
+  pending_ = acting_set_.size() - 1;
```

Comment:
comma initializer list?

Reply:
But note that then I will not be able to follow your previous suggestion above

Reply:
No, if it's explicit in the initializer list, it overrides the other initializer.